cirq-core/cirq/sim/simulator_base.py

```diff
@@ -291,12 +291,19 @@ def _run(
             for k, r in step_result._classical_data.records.items():
                 if k not in records:
                     records[k] = []
-                records[k].append(r)
+                records[k].append(np.array(r, dtype=np.uint8))
             for k, cr in step_result._classical_data.channel_records.items():
                 if k not in records:
                     records[k] = []
-                records[k].append([cr])
-        return {str(k): np.array(v, dtype=np.uint8) for k, v in records.items()}
+                records[k].append(np.array([cr], dtype=np.uint8))
+
+        def pad_evenly(results: Sequence[np.ndarray]):
+            largest = max([result.shape[0] for result in results])
+            return np.array(
+                [np.pad(result, ((0, largest - result.shape[0]), (0, 0))) for result in results]
+            )
+
+        return {str(k): pad_evenly(v) for k, v in records.items()}
```

Comment:
Can we tighten this up a bit to track the largest shapes as `records` is being constructed ? Also can we tighten up the convert statement somehow here ? It looks like we know the shape of the final array so perhaps we can do something like:

```python3
ret = np.zeros(final_shape)
for result in results:
    ret[<fancy_broadcast_stuff or masking_stuff>] =  result
return ret
```

Reply:
Good call, this was 10x slower than existing code for homogeneous arrays. New approach is 10x faster, and equivalent to old approach for homogeneous arrays.